### PR TITLE
bt-220: fix navigation in sign-up and root paths

### LINKED ISSUE #220: Connect SignUp to API
Connect UI to API

User story
As a Talent, I want to Sign Up for BSA hiring platform, so that can create my personal account on the platform for looking for a job

Acceptance criteria:

- If the fields Email and Password are empty, these fields light red color, and tips with information that this field is required appear, the user stays on this page
- After entering invalid data (does not match validation requirements) the notification about an invalid Email or Password appears on the screen, these fields light red color, and the user stays on this page
- If the user enters  Email and Password that was before used, the notification “User exists” appears, the user stays on this page
- After clicking on the button  “Continue” the user who chooses “I’m looking for a job” is redirected to the step “01/05” page of Onboarding Talent Flow
- Clicks on the button "I already have an account” redirects to the page "Sign In"

User story
As an Employer, I want to Sign Up for BSA hiring platform, so that can create my personal account on the platform to look a employees

- After clicking on the button  “Continue” the user who chooses “I’m hiring” is redirected to the Onboarding  Employer Flow
- Other points are the same as Talent’s role

### LINKED ISSUE #305: [BUG] User is redirected to the onboarding steps when opening the localhost
**Description:**
User is redirected to the onboarding steps when opening the localhost

**Environment:**
Windows 11 
Google Chrome 116.0

**Steps to reproduce:**

1. Open local host
2. Pay attention to the URL link and page

**Actual Result:**
User is redirected to the onboarding steps when opening the localhost

**Expected Result:**
User is redirected to the Sign in page when opening the localhost. User can't move to the onboarding without using a token

**Visual Proof (screenshots, videos, text):**
Screenshot_5

**Severity:**
Major (S3)

### PR DESCRIPTION
**Pull request by savinaDaria**
PR for :
- Creating navigation from Sign-up to Talent onboarding
- Setting Sign In as the default path, but in case user has token now he will be on the talent-onboarding path
- Change structure of index.tsx
- Also these talent-onboarding navigation problem fixed here
https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/issues/224#issuecomment-1711218718
https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/issues/224#issuecomment-1711714216

### CODE REVIEW

File: frontend/src/bundles/auth/pages/auth.tsx [hunk 1]
```diff
@@ -18,7 +21,17 @@ import { actions as authActions } from '../store/auth.js';
...
     const dispatch = useAppDispatch();
     const { pathname } = useLocation();
+    const navigate = useNavigate();
 
+    const { dataStatus } = useAppSelector(({ auth }) => ({
+        dataStatus: auth.dataStatus,
+    }));
+
+    useEffect(() => {
+        if (dataStatus === 'fulfilled') {
```
Comment (nikita-remeslov) [inline frontend/src/bundles/auth/pages/auth.tsx]: we should check if we have use in the store. Check protect-route logic

File: frontend/src/app/app.tsx
Comment (nikita-remeslov) [inline frontend/src/app/app.tsx:1]: we also should show loader until user load

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: frontend/src/app/app.tsx [hunk 2]
```diff
@@ -1,15 +1,44 @@
...
+        dataStatus: auth.dataStatus,
+    }));
+
+    const [isUserFullfilled, setIsUserFullfilled] = useState(false);
+
+    useEffect(() => {
+        setIsUserFullfilled(
+            dataStatus === 'fulfilled' || dataStatus == 'rejected',
+        );
+    }, [dataStatus]);
```
Comment (anton-otroshchenko) [inline frontend/src/app/app.tsx]: There is no need for this state

```
if(dataStatus !== ‘fulfilled’ || dataStatus !== ‘rejected’){
return <Loader />
}
```

and then return the rest

**Review by anton-otroshchenko — CHANGES_REQUESTED** [review]

**Review by JosuerBague — DISMISSED** [review]
Protected route redirect works.
Redirect after sign-in and sign-up works.

One issue found, when token is available and user is sign-in. Going to sign-in route loads sign-in and does not redirect to talent onboarding.

Will approve for demo.

File: frontend/src/app/app.tsx [hunk 3]
```diff
@@ -1,15 +1,35 @@
...
-            <PageLayout avatarUrl="" isOnline>
-                <RouterOutlet />
-            </PageLayout>
+            {dataStatus == 'fulfilled' || dataStatus == 'rejected' ? (
+                <PageLayout avatarUrl="" isOnline>
+                    <RouterOutlet />
+                </PageLayout>
+            ) : (
+                <Loader />
+            )}
```
Comment (nikita-remeslov) [inline frontend/src/app/app.tsx]:
```
if (dataStatus !== ...) {
  return <Loader />
}

return <Route />
```

Comment (nikita-remeslov) [reply frontend/src/app/app.tsx]: also, we should use enum for status.

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

**Comment by savinaDaria** [comment]
will try to figure out navigation after manually /sign-in /sign-up was printed in url path
now it doesn't loadCurrent user

**Comment by savinaDaria** [comment]
also changed children of App, because it broke my fixes

File: frontend/src/index.tsx [hunk 4]
```diff
@@ -36,51 +38,70 @@ createRoot(document.querySelector('#root') as HTMLElement).render(
...
-                                                            STEP_ROUTES.STEP_01,
-                                                        )}
-                                                    />
+                                                    <ProtectedRoute>
+                                                        <Navigate
+                                                            to={getStepRoute(
+                                                                STEP_ROUTES.STEP_01,
+                                                            )}
+                                                        />
+                                                    </ProtectedRoute>
```
Comment (JosuerBague) [inline frontend/src/index.tsx]:
```suggestion
                                                        <Navigate
                                                            to={getStepRoute(
                                                                STEP_ROUTES.STEP_01,
                                                            )}
                                                        />
```

File: frontend/src/index.tsx [hunk 5]
```diff
@@ -36,51 +38,70 @@ createRoot(document.querySelector('#root') as HTMLElement).render(
...
                                                 ),
                                             },
                                             {
                                                 path: AppRoute.TALENT_STEP,
-                                                element: <StepNavigation />,
+                                                element: (
+                                                    <ProtectedRoute>
+                                                        <StepNavigation />
+                                                    </ProtectedRoute>
+                                                ),
```
Comment (JosuerBague) [inline frontend/src/index.tsx]:
```suggestion
                                                        <StepNavigation />
                                                ),
```

File: frontend/src/app/app.tsx [hunk 6]
```diff
@@ -1,15 +1,36 @@
...
+
+    useEffect(() => {
+        void dispatch(authActions.loadUser());
+    }, [dispatch]);
+
+    if (
+        dataStatus !== DataStatus.FULFILLED &&
+        dataStatus !== DataStatus.REJECTED
+    ) {
+        return <Loader />;
```
Comment (JosuerBague) [inline frontend/src/app/app.tsx:26]: I am not sure if this should be wrapped in a modal. Perhaps in another ticket?

**Review by JosuerBague — CHANGES_REQUESTED** [review]

**Comment by savinaDaria** [comment]
also these navigation problem fixed here
https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/issues/224#issuecomment-1711218718
https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/issues/224#issuecomment-1711714216

**Review by nika-sirak — CHANGES_REQUESTED** [review]
When I click Continue page refreshes and only after that notification is rendering.
Another problem I recorded.

https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/assets/60103912/65f13982-ff41-4e14-b0fd-5c07bfdddd64

**Comment by JosuerBague** [comment]
I'm getting a flash of the Sign-In page after loading and before routing to onboarding steps.

https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/assets/90120572/40591496-5bc5-4624-80eb-4dd7746d75fd

File: frontend/src/index.tsx
Comment (nikita-remeslov) [inline frontend/src/index.tsx:1]: all these routes should be children of the App - https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/blob/development/frontend/src/index.tsx#L35-L79

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: frontend/src/app/app.tsx (hunk 6 above)
Comment (nikita-remeslov) [reply frontend/src/app/app.tsx:26]: No modal, just Loader

File: frontend/src/bundles/common/enums/app-route.enum.ts [hunk 7]
```diff
@@ -1,10 +1,13 @@
 const AppRoute = {
     ROOT: '/',
+    CANDIDATES: '/candidates',
+    CHATS: '/chats',
     SIGN_IN: '/sign-in',
     SIGN_UP: '/sign-up',
     RESET_PASSWORD: '/reset-password',
-    TALENT: '/talent',
+    TALENT: '/talent/onboarding/step',
```
Comment (nikita-remeslov) [inline frontend/src/bundles/common/enums/app-route.enum.ts]:
```suggestion
```

File: frontend/src/index.tsx [hunk 8]
```diff
@@ -38,46 +41,82 @@ createRoot(document.querySelector('#root') as HTMLElement).render(
...
                                             <Navigate
                                                 to={getStepRoute(
-                                                    STEP_ROUTES.STEP_01,
+                                                    StepsRoute.STEP_01,
                                                 )}
                                             />
                                         ),
                                     },
                                     {
                                         path: AppRoute.TALENT,
```
Comment (nikita-remeslov) [inline frontend/src/index.tsx]:
```suggestion
                                        path: AppRoute.TALENT_STEP,
```

File: frontend/src/index.tsx [hunk 9]
```diff
@@ -38,46 +41,82 @@ createRoot(document.querySelector('#root') as HTMLElement).render(
...
-                                                element: (
-                                                    <Navigate
-                                                        to={getStepRoute(
-                                                            STEP_ROUTES.STEP_01,
-                                                        )}
-                                                    />
-                                                ),
-                                            },
                                             {
                                                 path: AppRoute.TALENT_STEP,
```
Comment (nikita-remeslov) [inline frontend/src/index.tsx]:
```suggestion
                                                path: '',
```

File: frontend/src/index.tsx [hunk 10]
```diff
@@ -38,46 +41,82 @@ createRoot(document.querySelector('#root') as HTMLElement).render(
...
+                                        ),
+                                    },
+                                    {
+                                        path: AppRoute.RESET_PASSWORD,
+                                        element: (
+                                            <PublicRoute>
+                                                <Auth />
+                                            </PublicRoute>
+                                        ),
+                                    },
```
Comment (nikita-remeslov) [inline frontend/src/index.tsx]: move it to the top

File: frontend/src/app/app.tsx [hunk 11]
```diff
@@ -1,15 +1,38 @@
...
+        dataStatus: auth.dataStatus,
+    }));
+
+    const token = localStorage.getItem('token');
+    useEffect(() => {
+        if (token) {
+            void dispatch(authActions.loadUser());
+        } else {
+            dispatch(authActions.setUserNotAuthenticated());
+        }
```
Comment (nikita-remeslov) [inline frontend/src/app/app.tsx]:
```suggestion
        }
```

File: frontend/src/bundles/auth/store/slice.ts [hunk 12]
```diff
@@ -19,7 +19,11 @@ const initialState: State = {
 const { reducer, actions, name } = createSlice({
     initialState,
     name: 'auth',
-    reducers: {},
+    reducers: {
+        setUserNotAuthenticated: (state) => {
```
Comment (nikita-remeslov) [inline frontend/src/bundles/auth/store/slice.ts]: we dont need this reducer

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]
in general looks good

**Review by nikita-remeslov — APPROVED** [review]

**Review by nika-sirak — APPROVED** [review]

**Review by anton-otroshchenko — APPROVED** [review]

**Pull request merged**